File: nuam-kasse/backend/app/schemas/overview.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, computed_field, field_serializer

from app.core.money import format_money
from app.models.cash_period import CashPeriodStatus
from app.models.category import CategoryType


class OverviewCashPeriodRead(BaseModel):
    id: int
    name: str
    status: CashPeriodStatus
    start_date: date
    end_date: date | None
    currency: str

    model_config = ConfigDict(from_attributes=True)


class OverviewSummary(BaseModel):
    cash_period: OverviewCashPeriodRead
    opening_amount: str
    spent_amount: str
    income_amount: str
    remaining_amount: str
    expense_count: int
    active_expense_count: int
    voided_expense_count: int


class OverviewCategorySummary(BaseModel):
    category_id: int
    category_name: str
    icon_key: str
    color_key: str
    category_type: CategoryType
    image_updated_at: datetime | None = None
    image_preview_path: str | None = Field(default=None, exclude=True)
    expense_count: int
    total_amount: str
    percentage_of_spending: str

    @computed_field
    @property
    def has_custom_image(self) -> bool:
        return bool(self.image_preview_path and self.image_updated_at)

    @computed_field
    @property
    def image_url(self) -> str | None:
        if not self.has_custom_image:
            return None
        version = self.image_updated_at.isoformat()
        return f"/api/v1/categories/{self.category_id}/image?v={version}"


class OverviewUserSummary(BaseModel):
    user_id: int
    display_name: str
    expense_count: int
    total_amount: str
    percentage_of_spending: str


class OverviewExpenseCategory(BaseModel):
    id: int
    name: str
    icon_key: str
    color_key: str
    parent_category_id: int | None
    category_type: CategoryType
    image_updated_at: datetime | None = None
    image_preview_path: str | None = Field(default=None, exclude=True)

    @computed_field
    @property
    def has_custom_image(self) -> bool:
        return bool(self.image_preview_path and self.image_updated_at)

    @computed_field
    @property
    def image_url(self) -> str | None:
        if not self.has_custom_image:
            return None
        version = self.image_updated_at.isoformat()
        return f"/api/v1/categories/{self.id}/image?v={version}"

    model_config = ConfigDict(from_attributes=True)


class OverviewExpenseUser(BaseModel):
    id: int
    display_name: str

    model_config = ConfigDict(from_attributes=True)


class OverviewExpenseRead(BaseModel):
    id: int
    cash_period_id: int
    category: OverviewExpenseCategory
    amount: Decimal
    transaction_type: CategoryType
    currency: str
    created_by: OverviewExpenseUser
    created_at: datetime
    is_voided: bool
    voided_at: datetime | None = None
    voided_by: OverviewExpenseUser | None = None
    void_reason: str | None = None

    @field_serializer("amount")
    def serialize_amount(self, value: Decimal) -> str:
        return format_money(value)

    model_config = ConfigDict(from_attributes=True)


class CashPeriodOverview(BaseModel):
    summary: OverviewSummary
    categories: list[OverviewCategorySummary]
    users: list[OverviewUserSummary]
    recent_expenses: list[OverviewExpenseRead]


class PaginatedOverviewExpenses(BaseModel):
    items: list[OverviewExpenseRead]
    total: int
    limit: int
    offset: int
    has_more: bool

File: nuam-kasse/frontend/src/types/overview.ts
import type { CashPeriodStatus } from "./cashPeriod";

export type OverviewCashPeriod = {
  id: number;
  name: string;
  status: CashPeriodStatus;
  start_date: string;
  end_date: string | null;
  currency: "THB";
};

export type OverviewSummary = {
  cash_period: OverviewCashPeriod;
  opening_amount: string;
  spent_amount: string;
  remaining_amount: string;
  expense_count: number;
  active_expense_count: number;
  voided_expense_count: number;
};

export type CategorySummary = {
  category_id: number;
  category_name: string;
  icon_key: string;
  color_key: string;
  has_custom_image?: boolean;
  image_url?: string | null;
  image_updated_at?: string | null;
  expense_count: number;
  total_amount: string;
  percentage_of_spending: string;
};

export type UserSummary = {
  user_id: number;
  display_name: string;
  expense_count: number;
  total_amount: string;
  percentage_of_spending: string;
};

export type OverviewExpenseUser = {
  id: number;
  display_name: string;
};

export type OverviewExpenseCategory = {
  id: number;
  name: string;
  icon_key: string;
  color_key: string;
  parent_category_id: number | null;
  has_custom_image?: boolean;
  image_url?: string | null;
  image_updated_at?: string | null;
};

export type OverviewExpense = {
  id: number;
  cash_period_id: number;
  category: OverviewExpenseCategory;
  amount: string;
  currency: "THB";
  created_by: OverviewExpenseUser;
  created_at: string;
  is_voided: boolean;
  voided_at: string | null;
  voided_by: OverviewExpenseUser | null;
  void_reason: string | null;
};

export type CashPeriodOverview = {
  summary: OverviewSummary;
  categories: CategorySummary[];
  users: UserSummary[];
  recent_expenses: OverviewExpense[];
};

export type OverviewExpenseSort = "created_at_desc" | "created_at_asc" | "amount_desc" | "amount_asc";

export type OverviewExpenseFilters = {
  category_id?: number;
  created_by_user_id?: number;
  date_from?: string;
  date_to?: string;
  include_voided?: boolean;
  limit?: number;
  offset?: number;
  sort?: OverviewExpenseSort;
};

export type PaginatedOverviewExpenses = {
  items: OverviewExpense[];
  total: number;
  limit: number;
  offset: number;
  has_more: boolean;
};
